orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.h

```diff
@@ -0,0 +1,118 @@
+ *         such as [..., batch_size, sequence_length, ...].
+ *     (2) Its 3nd input is a scalar constant initializer which is the padding idx that should >= 0.
+ * 2. Iterate the subgraph that starts from the Embedding node and find all inputs and outputs of the subgraph.
+ * 3. Insert pattern of 'Reshape + ShrunkenGather' before each input of the subgraph to eliminate the padding.
+ * 4. Insert pattern of 'GatherGrad + Reshape' after each output of the subgraph to restore the original shape.
+ *    This is needed to ensure not to affect subsequent computations
+ *
+ * For example, given the following graph:
+ *
+ *              input_ids [batch_size, seq_length]   padding_idx [1]
+ *       \                   \                               /                /
+ *         \                  \                             /                /
+ *           \                 \                           /                /
+ *                                   Aten:embedding
+ *                                         |
+ *                                         |
+ *                     input               |
+ *                             \
+ *                                     Subgraph
+ *
+ *                                         |
+ *                                       output
+ *
+ *
+ * After the transformation:
+ *
+ *             input_ids [batch_size, seq_length]
+ *                |      \
+ *                |       \        [-1]     padding_idx [1]
+ *                |        \       /         /
+ *                |         Reshape         /
+ *                |              \         /
+ *                |                  Sub
+ *                |    [-1]           |
+ *                |    /            NonZero
+ *               Reshape              |
+ *                     \            Squeeze
+ *                      \            /
+ *                       \          /(valid_token_index)
+ *                        \        /
+ *                      ShrunkenGather
+ *                              \
+ *      \                        \  shape:[valid_token]                padding_idx [1]        /
+ *       \                        \                                     /                    /
+ *        \                        \                                   /                    /
+ *                                          Aten:embedding
+ *                                                  |
+ *                                                  |
+ *       input [batch_size, seq_length]             |
+ *           \                                      |
+ *            \     [-1]                            |
+ *             \    /                               |
+ *             Reshape       (valid_token_index)    |
+ *                  \           /                   |
+ *                   ShrunkenGatherGrad             |  shape:[valid_token]
```

Comment:
```suggestion
 *                   ShrunkenGatherGrad             |  shape:[valid_token]
```

should not this be "ShrunkenGather"?